[RFC] Use module pattern with vim/shared.lua

h-michael opened the pull request:
It's a bit cumbersome for us to add an export target every time we define a new function.
And it's also cumbersome to care about the order of definition when creating a new function by referring to other functions in the module.

ref: #10724

h-michael committed:
runtime: Use module pattern with vim/shared.lua

It's a bit cumbersome for us to add an export target every time we define a new function.
It's also cumbersome to care about the order of definition when creating a new function by referring to other functions in the module.

justinmk commented:
+1 but we need to fix the lua2dox.lua script to handle this. It is used by this script:

    scripts/gen_vimdpc.py

h-michael commented:
@justinmk Thanks :) I'll see #9740 later.

justinmk commented on runtime/lua/vim/shared.lua:
```diff
@@ -4,32 +4,32 @@
+  }
+
   return deepcopy_funcs[type(orig)](orig)
 end
```
could use an extra closure, to avoid re-defining `_id` and `deepcopy_funcs` at each call:

```lua
module.deepcopy = function()
  local function _id(v)
    return v
  end

  local deepcopy_funcs = {
    table = function(_orig)
      local copy = {}
      for k, v in pairs(_orig) do
        copy[module.deepcopy(k)] = module.deepcopy(v)
      end
      return copy
    end,
    number = _id,
    string = _id,
    ['nil'] = _id,
    boolean = _id,
  }

  return function(orig)
    return deepcopy_funcs[type(orig)](orig)
  end
end()
```

h-michael replied:
This has syntax error. Lua may not use () for immediate execution after end.

justinmk replied:
just wrap it in parens. see `paste` for example.

h-michael replied:
I see :)

justinmk commented on runtime/lua/vim/shared.lua:
```diff
@@ -92,8 +92,8 @@ end
 --@returns List-like table of the split components.
-local function split(s,sep,plain)
-  local t={} for c in gsplit(s, sep, plain) do table.insert(t,c) end
+module.split = function(s,sep,plain)
```
we use this style (where possible):

    function module.split(...)

justinmk commented:
> @justinmk Thanks :) I'll see #9740 later.

BTW: if it makes things easier, we can use the module's name instead of `module.` for the prefix.  (as @blueyed proposed  :)

justinmk commented on runtime/lua/vim/shared.lua:
```diff
@@ -4,34 +4,36 @@
 -- test-suite. If, in the future, Nvim itself is used to run the test-suite
 -- instead of "vanilla Lua", these functions could move to src/nvim/lua/vim.lua
 
+local shared = {}
```
shared.lua is a special case, we actually consider it part of the `vim` module. So I think this should be `vim` -- that will help ctags, for example.

We won't ever need to `require('vim')` in this module (because this module, by definition, is pure Lua--without Nvim), so there won't be any conflict here.

h-michael replied:
OK. BTW when I first read Neovim code, I was a little confused by the file name shared.lua.
Although it is a module shared with test helper, util.lua seems to be easier to understand, what do you think?

justinmk replied:
why would that be easier to understand?

h-michael replied:
Maybe because of my little experience.

justinmk replied:
> But that module test file nema is `utility_funcitons_spec.lua`.

Yeah, I don't know what that test is named so. We should rename it to `vim_spec.lua`

justinmk commented on runtime/lua/vim/shared.lua:
```diff
@@ -4,34 +4,36 @@
 --@returns New table of copied keys and (nested) values.
-local function deepcopy(orig)
-  error(orig)
-end
```
I think we still need a "dummy" definition. When I ran `gen_vimdoc.py` from this PR, `deepcopy` was missing from `runtime/doc/if_lua.txt`.

```lua
function shared.deepcopy(orig) end  -- luacheck: no unused
```

h-michael committed:
scripts/lua2dox.lua: Remove class declaration block

Judging class definitions in the form "string.find (line, '=% s * class% (')"
must force writing class definitions in this format, but such a mechanism is Absent.
Also, Lua has no formal class in the language specification, and implements inheritance with setmetadable.
To detect this, we should have a parser for it, not a simple regular expression.
